waiter/src/waiter/token.clj

```diff
@@ -159,40 +167,43 @@
 
    If handling POST, validates that the user is the creator of the token if it already exists.
    Then, updates the configuration for the token in the database using the newest password."
-  [synchronize-fn kv-store waiter-hostname can-run-as? make-peer-requests-fn validate-service-description-fn {:keys [request-method] :as req}]
+  [clock synchronize-fn kv-store waiter-hostname entitlement-manager make-peer-requests-fn validate-service-description-fn
+   {:keys [headers request-method] :as req}]
   ;;if post, validate that this is a valid job schema & that the user == the kerberos user, then sign & store in riak
   ;;  remember that we need an extra field in this schema, which is who is allowed to use this. Could be "*" or a string username
   ;;if get, return whatever data's in riak
   (case request-method
     :delete (try
-              (let [req-headers (:headers req)
-                    {:keys [token]} (sd/retrieve-token-from-service-description-or-hostname req-headers req-headers waiter-hostname)
-                    current-user (get req :authorization/user)]
+              (let [{:keys [token]} (sd/retrieve-token-from-service-description-or-hostname headers headers waiter-hostname)
+                    authenticated-user (get req :authorization/user)
+                    request-params (:query-params (ring-params/params-request req))
+                    excise (utils/request-flag request-params "excise")]
                 (if token
-                  (let [{:keys [service-description-template token-metadata]} (sd/token->token-description kv-store token)]
+                  (let [{:keys [service-description-template token-metadata]} (sd/token->token-description kv-store token :show-deleted excise)]
                     (if (and service-description-template (not-empty service-description-template))
                       (let [token-owner (get token-metadata "owner")]
-                        (when-not (can-run-as? current-user token-owner)
-                          (throw (ex-info "User not allowed to delete token"
-                                          {:existing-owner token-owner
-                                           :current-user current-user
-                                           :status 403})))
-                        (delete-service-description-for-token synchronize-fn kv-store token token-owner)
+                        (if excise
+                          (when-not (authz/sync-token? entitlement-manager authenticated-user token token-metadata)
```

Comment:
Should `sync-token?` be renamed also.  Syncing is external to Waiter.  Waiter decides whether to allow "system" level access, i.e. ignoring normal permission scheme.